tracer/src/Datadog.Trace/Debugger/Instrumentation/AsyncMethodDebuggerInvoker.cs

```diff
@@ -0,0 +1,371 @@
+            asyncState.SnapshotCreator.EndEntry(hasArgumentsOrLocals);
+            return asyncState;
+        }
+
+        /// <summary>
+        /// Logs the given <paramref name="arg"/> ByRef.
+        /// </summary>
+        /// <typeparam name="TArg">Type of argument.</typeparam>
+        /// <param name="arg">The argument to be logged.</param>
+        /// <param name="type">The type of the argument</param>
+        /// <param name="paramName">The argument name</param>
+        /// <param name="index">Index of given argument.</param>
+        /// <param name="asyncState">Debugger asyncState</param>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static void LogArg<TArg>(ref TArg arg, Type type, string paramName, int index, ref AsyncMethodDebuggerState asyncState)
+        {
+            asyncState.SnapshotCreator.CaptureArgument(arg, paramName, index == 0, asyncState.HasLocalsOrReturnValue, type);
+            asyncState.HasLocalsOrReturnValue = false;
+            asyncState.HasArguments = true;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static void LogLocal<TLocal>(ref TLocal local, Type type, string localName, int index, ref AsyncMethodDebuggerState asyncState)
+        {
+            asyncState.SnapshotCreator.CaptureLocal(local, localName, index == 0 && !asyncState.HasLocalsOrReturnValue, type);
+            asyncState.HasLocalsOrReturnValue = true;
+            asyncState.HasArguments = false;
+        }
+
+        /// <summary>
+        /// Logs the given <paramref name="local"/> ByRef.
+        /// </summary>
+        /// <typeparam name="TLocal">Type of local.</typeparam>
+        /// <param name="local">The local to be logged.</param>
+        /// <param name="index">index of given argument.</param>
+        /// <param name="asyncState">Debugger asyncState</param>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static void LogLocal<TLocal>(ref TLocal local, int index, ref AsyncMethodDebuggerState asyncState)
+        {
+            if (!asyncState.IsActive)
+            {
+                return;
+            }
+
+            var localNamesFromPdb = asyncState.MethodMetadataInfo.LocalVariableNames;
+            if (!MethodDebuggerInvoker.TryGetLocalName(index, localNamesFromPdb, out var localName))
+            {
+                return;
+            }
+
+            LogLocal(ref local, typeof(TLocal), localName, index, ref asyncState);
+        }
+
+        /// <summary>
+        /// End Method with Return value invoker - the MoveNext method always return void but here we send the kick-of method return value
+        /// This method is called from either (1) the outer-most catch clause when the async method threw exception
+        /// or (2) when the async method has logically ended.
+        /// In this phase we have the correct async context in hand because we already set it in the BeginMethod.
```

Comment:
```suggestion
        /// End Method with Return value invoker - the MoveNext method always returns void but here we send the kick-off method's return value
        /// This method is called from either (1) the outer-most catch clause (when the async method threw exception)
        /// or (2) when the async method has logically ended.
        /// In this phase we have the correct async context in hand because we already set it in the BeginMethod.
```

Reply:
Fixed in 5a103fbe57cedbc9e13568d4372c82a921c46e26